### DISCUSSION TITLE
Deno feature?

### DISCUSSION THREAD

**Discussion by jcbhmr** [devcontainers-community/features#18] [Ideas]
from https://github.com/devcontainers-community/templates/discussions/24#discussioncomment-5773509 thread

> I agree with adding Deno.
> 
> Perhaps we also need a Feature to install Deno....... (Deno does not have an official arm64 binary, so the Feature may need an option to install an unofficial arm64 binary)

> @devcontainers-contrib does actually already have a deno feature! in fact, there's two! it's just that the way its gone about is... not not the most _efficient_ (install asdf just to install deno?) or "official"ish imo. but that's my opinion.
> 
> ![image](https://user-images.githubusercontent.com/61068799/235495495-18e19a06-cd2a-4af0-a815-d5e5de4985c5.png)
> 
> we **could** use that or maintain our own that just relegates to installing curl, unzip, and then `curl -fsSL https://deno.land/x/install/install.sh | sh` depending on whether or not that's worth the effort to make.
> 
> > so the Feature may need an option to install an unofficial arm64 binary)
> 
> this would be a **very valuable** addon/expansion/option! 🥇 i dont think the current two features offer that

> I do not know if the official installation scripts match the demand to install a specific version. (I've heard that Deno itself has the ability to switch versions, so this may not actually be a problem.)

> yup! you can switch version on-the-fly via the `deno upgrade` command:
> 
> ![image](https://user-images.githubusercontent.com/61068799/235499392-a272b73f-2066-4603-84da-15d3de47afd4.png)
> 
> https://deno.com/manual@v1.33.1/getting_started/installation#updating
> 
> looks like you can also set the version via a script arg `$1` of the installer
> 
> ```shell
> curl -fsSL https://deno.land/x/install/install.sh | sh -s v1.24.1
> ```

> > looks like you can also set the version via a script arg `$1` of the installer
> > ```shell
> > curl -fsSL https://deno.land/x/install/install.sh | sh -s v1.24.1
> > ```
> 
> Sounds good! The Julia Feature does the same thing. https://github.com/JuliaLang/devcontainer-features/blob/45c7a4d2c7ae4f451ddc1692705e15ce931c43dc/src/julia/install.sh#L52-L58
> 
> As for arm64 support, I think we need to write a complete installation script for now.

cc @eitsupi

**Comment by jcbhmr** [comment]
someone else has done something cool with that arm64 deno thing! 😲 https://github.com/LukeChannings/deno-arm64 👍 very cool project @LukeChannings! thanks for making arm64 builds of deno for everyone! 🤗

**Reply by jcbhmr** [reply]
only thing we'd want to be AWARE OF is that this is one guy. not a guarenteed upkeep of arm64 binaries... also have to make sure we trust them to not yank that gist installer by accidentally deleting it during a spring cleaning session

so as @eitsupi mentioned,

> As for arm64 support, I think we need to write a complete installation script for now.

probably, yeah 😔

**Reply by LukeChannings** [reply]
👋, I have automated this project now, so I usually have a build 30 minutes or so after the official version.

Sometimes they break ARM builds, and that usually takes a day or two to merge a fix for - I tend to patch the broken release, so there isn't a missing version.

I have patched the upgrade function for my binaries, so ` deno upgrade` works as expected. When deno officially has ARM releases (they don't really seem interested...), I'll make a final release that swaps the upgrade URL back to the official URL, and then maybe do a few more releases to get everyone switched over. 

Regarding the install script, it's a small modification to the official installer, and you can (and should) pin it to a git hash.

**Reply by jcbhmr** [reply]
> 👋, I have automated this project now, so I usually have a build 30 minutes or so after the official version.

that's awesome! 🚀 that makes me feel alot better about trusting you to keep things moving smoothly for future releases without getting stuck on vX.Y.Z forever! ❤️

btw thanks for volunteering your time to do this stuff! really makes it easy for Raspberry Pi or other ARM linux pcs to run deno. 💕

> I have patched the upgrade function for my binaries, so  deno upgrade works as expected. When deno officially has ARM releases (they don't really seem interested...), I'll make a final release that swaps the upgrade URL back to the official URL, and then maybe do a few more releases to get everyone switched over.

😱 and that's even **more awesome-er!**
_"(they don't really seem interested...)"_ 🙁

> Regarding the install script, it's a small modification to the official installer, and you can (and should) pin it to a git hash.

👍

**Comment by jcbhmr** [comment]
we can pin a version of the gist https://gist.github.com/LukeChannings/09d53f5c364391042186518c8598b85e as recommended in the readme
its a MODIFICATION of the original install.sh script that ADDS support for arm64, so

```sh
curl -s https://gist.githubusercontent.com/LukeChannings/09d53f5c364391042186518c8598b85e/raw/ac8cd8c675b985edd4b3e16df63ffef14d1f0e24/deno_install.sh | sh
```

is in theory just an addon. only issue is it's now one step out-of-sync with the main deno_install official deno project if they do any changes (which i admit is unlikely)
